pkg/virt-handler/selinux/context_executor.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package selinux
+
+import (
+	"fmt"
+	"os/exec"
+	"runtime"
+	"syscall"
+
+	"github.com/opencontainers/selinux/go-selinux"
+)
+
+type SELinuxContextExecutor struct {
+	cmdToExecute  *exec.Cmd
+	desiredLabel  string
+	originalLabel string
+	pid           int
+}
+
+func NewSELinuxContextExecutor(pid int, cmd *exec.Cmd) (*SELinuxContextExecutor, error) {
+	desiredLabel, err := getSelinuxLabelForPID(pid)
+	if err != nil {
+		return nil, err
+	}
+	originalLabel, err := getOriginalSelinuxLabel()
+	if err != nil {
+		return nil, err
+	}
+	return &SELinuxContextExecutor{
+		pid:           pid,
+		cmdToExecute:  cmd,
+		desiredLabel:  desiredLabel,
+		originalLabel: originalLabel,
+	}, nil
+}
+
+func (ce SELinuxContextExecutor) Execute() error {
+	if isSELinuxEnabled() {
+		if err := ce.setDesiredContext(); err != nil {
+			return err
+		}
+	}
+	defer ce.resetContext()
+
+	preventFDLeakOntoChild()
+	if err := ce.cmdToExecute.Run(); err != nil {
+		return fmt.Errorf("failed to execute command in launcher namespace %d: %v", ce.pid, err)
+	}
+	return nil
+}
+
+func (ce SELinuxContextExecutor) setDesiredContext() error {
+	runtime.LockOSThread()
+	if err := selinux.SetExecLabel(ce.desiredLabel); err != nil {
+		return fmt.Errorf("failed to switch selinux context to %s. Reason: %v", ce.desiredLabel, err)
+	}
+	return nil
+}
+
+func (ce SELinuxContextExecutor) resetContext() error {
+	err := selinux.SetExecLabel(ce.originalLabel)
+	runtime.UnlockOSThread()
+	return err
+}
+
+func isSELinuxEnabled() bool {
+	_, selinuxEnabled, err := NewSELinux()
+	return err == nil && selinuxEnabled
+}
+
+func getSelinuxLabelForPID(pid int) (string, error) {
+	launcherSELinuxLabel, err := selinux.FileLabel(fmt.Sprintf("/proc/%d/attr/current", pid))
+	if err != nil {
+		return "", fmt.Errorf("could not retrieve pid %d selinux label: %v", pid, err)
+	}
+	return launcherSELinuxLabel, nil
+}
+
+func getOriginalSelinuxLabel() (string, error) {
+	launcherSELinuxLabel, err := selinux.FileLabel("/proc/thread-self/attr/current")
+	if err != nil {
+		return "", fmt.Errorf("could not retrieve original pid selinux label: %v", err)
+	}
+	return launcherSELinuxLabel, nil
+}
+
+func preventFDLeakOntoChild() {
+	// we want to share the parent process std{in|out|err}
```

Comment:
Don't we want to **not** share it? It will be nice to place it as a doctext instead.

Reply:
Hm, this is a private function. I don't think it makes sense to add doctext on something private to the pkg.

And no, we *do* want to share these FDs w/ the parent process.

Reply:
I do not understand something: When an exec occurs, the fd/s are getting closed in order not to pass them to the child, no? I guess I am missing something.

Reply:
> I do not understand something: When an exec occurs, the fd/s are getting closed in order not to pass them to the child, no? I guess I am missing something.

That does not happen, and afaiu, is the sole reasoning for the cloexec flag.

Taken from https://www.gnu.org/software/libc/manual/html_node/Descriptor-Flags.html

"""
Macro: int FD_CLOEXEC
This flag specifies that the file descriptor should be closed when an exec function is invoked; see Executing a File. When a file descriptor is allocated (as with open or dup), this **bit is initially cleared on the new file descriptor, meaning that descriptor will survive into the new program after exec**.
"""

A couple of interesting links about this can be found in the commit that introduced this: https://github.com/kubevirt/kubevirt/commit/b2e6ad49490876eca47dd0a09c539012fc5d5798

Reply:
I think we are saying the same thing. The problem is that the comment is focusing on the dirst 3 fd/s, while the execution handles the rest.
I would comment on the function as doctext and explain what it does: Keeps the stdin, stdout and stderr open and shared and the rest are closed on the child.

Reply:
> I would comment on the function as doctext and explain what it does: Keeps the stdin, stdout and stderr open and shared and the rest are closed on the child.

Tried to shed some more light into this in the already existent comment. 

Having said that, I'm not going to document private functions. That makes no sense.